ddcommon-net/src/http.rs

```diff
@@ -0,0 +1,287 @@
+                hex::encode(path.as_os_str().as_bytes())
+            }
+            #[cfg(not(unix))]
+            {
+                hex::encode(path.to_string_lossy())
+            }
+        };
+        Uri::builder()
+            .scheme(scheme)
+            .authority(hex_encoded_path)
+            .build()
+    }
+
+    #[cfg(not(unix))]
+    fn from_unix_path(path: &path::Path) -> http::Result<Uri> {
+        Uri::builder()
+            .scheme("unix")
+            .authority(path.to_string_lossy())
+            .build()
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error(transparent)]
+    Dns(#[from] rustls::pki_types::InvalidDnsNameError),
+
+    #[error(transparent)]
+    Http(#[from] http::Error),
+
+    #[error(transparent)]
+    Hyper(#[from] hyper::Error),
+
+    #[error(transparent)]
+    Io(#[from] io::Error),
+
+    #[error(transparent)]
+    Rustls(#[from] rustls::Error),
+
+    #[error(transparent)]
+    Timeout(#[from] Elapsed),
+
+    #[error("unsupported scheme: `{0}`")]
+    UnsupportedScheme(String),
+
+    #[error("user requested cancellation")]
+    UserRequestedCancellation,
+}
+
+/// Sends a blocking HTTP request using the provided runtime, inferring the
+/// connector type to use from the request's URI scheme. This is the primary
+/// API for sending an infrequent HTTP request, such as once per minute for
+/// profilers.
+pub fn one_shot<B>(
+    runtime: &tokio::runtime::Runtime,
+    request: Request<B>,
+    cancel: Option<&CancellationToken>,
+    timeout: Option<time::Duration>,
+) -> StdResult<Response<Incoming>, Error>
+where
+    B: Body + Send + 'static,
+    B::Data: Send,
+    B::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
+    runtime.block_on(async move {
+        tokio::select! {
+            result = async {
+                Ok(match timeout {
+                    Some(t) => tokio::time::timeout(t, send_and_infer_connector(request)).await?,
+                    None => send_and_infer_connector(request).await,
+                }?)}
+            => result,
+            _ = async { match cancel {
+                    Some(token) => token.cancelled().await,
+                    // If no token is provided, future::pending() provides a no-op future that never resolves
+                    None => std::future::pending().await,
+                }}
+            => Err(Error::UserRequestedCancellation),
+        }
+    })
+}
+
+pub async fn send_and_infer_connector<B>(
+    request: Request<B>,
+) -> StdResult<Response<Incoming>, Error>
+where
+    B: Body + Send + 'static,
+    B::Data: Send,
+    B::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
+    let uri = request.uri();
+    match uri.scheme() {
+        None => Err(Error::UnsupportedScheme(String::new())),
+        Some(scheme) => match scheme.as_str() {
+            "http" => send_http(request).await,
+            "https" => send_https(request).await,
+            #[cfg(unix)]
```

Comment:
Do we eventually want to support this on Windows? https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/